OpenRA.Mods.Common/AI/States/NavyStates.cs

```diff
@@ -0,0 +1,194 @@
+			if (!owner.IsValid)
+				return null;
+
+			// For ships, cheat and move towards enemy naval production, if any.
+			var t = owner.World.ActorsHavingTrait<Building>().Where(a
+				=> owner.Bot.Info.BuildingCommonNames.NavalProduction.Contains(a.Info.Name)
+				&& a.AppearsHostileTo(owner.Units.First())).FirstOrDefault();
+
+			// If naval yard is too far away, return it.
+			// Else, FindClosest below will find suitable enemy targets :)
+			if (t != null && (t.Location - owner.Units.First().Location).LengthSquared
+					> owner.Bot.Info.MaxBaseRadius * owner.Bot.Info.MaxBaseRadius)
+				return t;
+
+			return owner.Bot.FindClosestEnemy(owner.Units.FirstOrDefault().CenterPosition);
+		}
+	}
+
+	class NavyUnitsIdleState : NavyStateBase, IState
+	{
+		public void Activate(Squad owner) { }
+
+		public void Tick(Squad owner)
+		{
+			if (!owner.IsValid)
+				return;
+
+			if (!owner.IsTargetValid)
+			{
+				var t = FindClosestEnemy(owner);
+				if (t == null) return;
+				owner.TargetActor = t;
+			}
+
+			var enemyUnits = owner.World.FindActorsInCircle(owner.TargetActor.CenterPosition, WDist.FromCells(10))
+				.Where(unit => owner.Bot.Player.Stances[unit.Owner] == Stance.Enemy).ToList();
+
+			if (enemyUnits.Any())
+			{
+				if (AttackOrFleeFuzzy.Default.CanAttack(owner.Units, enemyUnits))
+				{
+					foreach (var u in owner.Units)
+						owner.Bot.QueueOrder(new Order("AttackMove", u, false) { TargetLocation = owner.TargetActor.Location });
+
+					// We have gathered sufficient units. Attack the nearest enemy unit.
+					owner.FuzzyStateMachine.ChangeState(owner, new NavyUnitsAttackMoveState(), true);
+					return;
+				}
+				else
+					owner.FuzzyStateMachine.ChangeState(owner, new NavyUnitsFleeState(), true);
+			}
+		}
+
+		public void Deactivate(Squad owner) { }
+	}
+
+	class NavyUnitsAttackMoveState : NavyStateBase, IState
+	{
+		public void Activate(Squad owner) { }
+
+		public void Tick(Squad owner)
+		{
+			if (!owner.IsValid)
+				return;
+
+			if (!owner.IsTargetValid)
+			{
+				var t = FindClosestEnemy(owner);
+				if (t != null)
+					owner.TargetActor = t;
+				else
+				{
+					owner.FuzzyStateMachine.ChangeState(owner, new NavyUnitsFleeState(), true);
+					return;
+				}
+			}
+
+			var leader = owner.Units.ClosestTo(owner.TargetActor.CenterPosition);
+			if (leader == null)
+				return;
+			var ownUnits = owner.World.FindActorsInCircle(leader.CenterPosition, WDist.FromCells(owner.Units.Count) / 3)
+				.Where(a => a.Owner == leader.Owner && owner.Units.Contains(a)).ToHashSet();
+			if (ownUnits.Count < owner.Units.Count)
+			{
+				owner.Bot.QueueOrder(new Order("Stop", leader, false));
+				foreach (var unit in owner.Units.Where(a => !ownUnits.Contains(a)))
+					owner.Bot.QueueOrder(new Order("AttackMove", unit, false) { TargetLocation = leader.Location });
+			}
+			else
+			{
+				var enemies = owner.World.FindActorsInCircle(leader.CenterPosition, WDist.FromCells(12))
+					.Where(a1 => !a1.Disposed && !a1.IsDead);
```

Comment:
While we're at it: `IsDead` contains an `Disposed` check, so you can drop the extra one here. (And also drop the `1` in the variable name, please.)